### PR TITLE
Connect difficulty manager to TheoryEngine

### PR DESCRIPTION
**Pull request by ajreynol**
This also introduces the `produceDifficulty` option which is analogous to `produceUnsatCores`.

It requires another unsat cores mode `PP_ONLY`, indicating that we are only tracking proofs of preprocessing.  This option should perhaps be renamed to `proofMode` instead of `unsatCoresMode`, since its use is more general than for unsat cores.  This will be addressed in a future refactoring.

### CODE REVIEW

File: src/theory/relevance_manager.h [hunk 1]
```diff
@@ -73,7 +79,7 @@ class RelevanceManager
   typedef context::CDList<Node> NodeList;
 
  public:
-  RelevanceManager(context::UserContext* userContext, Valuation val);
+  RelevanceManager(Env& env, Valuation val);
```
Comment (nafur) [inline src/theory/relevance_manager.h:82]: Should we make `RelevanceManager` an `EnvObj`?

Comment (ajreynol) [reply src/theory/relevance_manager.h:82]: Actually this is unecessary for now, I'll revert this.

**Comment by HanielB** [comment]
Regarding the option, an issue with that change is that there would be unsat-core-specific stuff in it, like whether we are using assumptions or not in the SAT solver. Maybe we should have two different options, one about how proofs are being generated and another about how unsat cores are being generated. A disadvantage with that is that it'd add more checks to `set_defaults`.

File: src/theory/relevance_manager.h [hunk 2]
```diff
@@ -157,6 +169,17 @@ class RelevanceManager
    * aborts and indicates that all literals are relevant.
    */
   bool d_success;
+  /** Are we tracking the sources of why a literal is relevant */
+  bool d_trackRSetExp;
+  /** Miniscope top-level AND */
```
Comment (HanielB) [inline src/theory/relevance_manager.h:174]: clarify

File: src/theory/relevance_manager.cpp [hunk 3]
```diff
@@ -64,8 +78,10 @@ void RelevanceManager::addAssertionsInternal(std::vector<Node>& toProcess)
   while (i < toProcess.size())
   {
     Node a = toProcess[i];
-    if (a.getKind() == AND)
+    if (d_miniscopeTopLevel && a.getKind() == AND)
     {
+      // difficulty tracking requires splitting top-level AND earlier
```
Comment (HanielB) [inline src/theory/relevance_manager.cpp:83]: clarify

**Review by HanielB — CHANGES_REQUESTED** [review]

**Review by HanielB — APPROVED** [review]

**Pull request merged**